src/main/scala/com/fulcrumgenomics/bam/Bams.scala

```diff
@@ -490,4 +492,35 @@ object Bams extends LazyLogging {
       None
     }
   }
+
+  /** Requires that the header is queryname sorted or query grouped. */
+  def requireTemplateGrouped(header: SAMFileHeader, toolName: String): Unit = {
+    require(header.getSortOrder == SortOrder.queryname || header.getGroupOrder == GroupOrder.query,
+      "Input was not queryname sorted or query grouped, found: " +
+        s"SO:${header.getSortOrder} GO:${header.getGroupOrder}" +
+        Option(header.getAttribute("SS")).map(ss => f" SS:$ss").getOrElse("") +
+        f". Use `samtools sort -n -u in.bam | fgbio $toolName -i /dev/stdin`"
+    )
+  }
+
+  /** Builds a [[Writer]] of [[SamRecord]]s that regenerates the NM, UQ, and MD tags using the given map of reference
+    * sequences.
+    *
+    * @param writer the writer to write to
+    * @param ref the path to the reference FASTA
+    */
+  def regenerateNmUqMdTagsWriter(writer: SamWriter, ref: PathToFasta): Writer[SamRecord] with Closeable = {
+    logger.info("Reading the reference fasta into memory")
+    val refMap = ReferenceSequenceIterator(ref, stripComments=true).map { ref => ref.getContigIndex -> ref}.toMap
+    logger.info(f"Read ${refMap.size}%,d contigs.")
```

Comment:
Can we bump these to debug?  I don't think logging statements like this really belong in a utility function, so I would prefer to remove or at least make debug.

Reply:
Ok.  It does take a while to load some large references, so it's nice to have to not duplicate logging messages everywhere. But here it makes more sense to move them into the tools